quinn/src/lib.rs

```diff
@@ -330,50 +316,63 @@ impl EndpointInner {
             buffered_incoming: VecDeque::new(),
             driver: None,
+            connecting: FnvHashMap::default(),
+            connections: FnvHashMap::default(),
+            sender,
+            events,
         }
     }
 
-    /// Wake up a blocked `Driver` task to process I/O
-    fn notify(&self) {
-        if let Some(x) = self.driver.as_ref() {
-            x.notify();
+    fn drive_connecting(&mut self) {
+        for new in self.connecting.values_mut() {
+            new.poll().unwrap();
         }
     }
 
-    fn drive_recv(
-        &mut self,
-        now: Instant,
-        wrapped: Rc<RefCell<EndpointInner>>,
-    ) -> Result<bool, io::Error> {
+    fn drive_recv(&mut self, now: Instant) -> Result<bool, io::Error> {
         let mut buf = [0; 64 * 1024];
         let mut recvd = 0;
         loop {
             match self.socket.poll_recv(&mut buf) {
                 Ok(Async::Ready((n, addr, ecn))) => {
-                    if let Some(ch) = self.inner.handle(now, addr, ecn, (&buf[0..n]).into()) {
-                        self.pending.insert(ch, Pending::new(None));
-                        match self.incoming.poll_ready() {
-                            Ok(Async::Ready(())) => {
-                                self.incoming
-                                    .start_send(NewConnection::new(wrapped.clone(), ch))
-                                    .unwrap();
-                                self.inner.accept();
+                    match self.inner.handle(now, addr, ecn, (&buf[0..n]).into()) {
+                        Some((handle, ConnectionEvent::NewConnection(conn))) => {
+                            let conn = self.create_connection(handle, conn, None);
+                            let conn = Rc::new(RefCell::new(conn));
+                            match self.incoming.poll_ready() {
+                                Ok(Async::Ready(())) => {
+                                    self.incoming
+                                        .start_send(NewConnection {
+                                            driver: ConnectionDriver(conn.clone()),
+                                            connection: Connection(conn.clone()),
+                                            incoming: IncomingStreams(conn),
+                                        })
+                                        .unwrap();
+                                    self.inner.accept();
+                                }
+                                _ => {
+                                    self.buffered_incoming.push_back((handle, Connection(conn)));
+                                }
                             }
-                            _ => {
-                                self.buffered_incoming.push_back(ch);
+                        }
+                        Some((ch, e)) => {
+                            if let Some(sender) = self.connections.get_mut(&ch) {
+                                sender.unbounded_send(e).unwrap();
```

Comment:
Future work, but one way to make this bounded would be to construct the bulk of the driver as a `Stream<Item=(ConnectionHandle, ConnectionEvent)>`, then forward that into a multiplexing sender. That would mean a hung connection = a hung endpoint, of course.

Reply:
Let's defer that for now.